photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/estimators/GaussianProcessEstimator.scala

```diff
@@ -0,0 +1,148 @@
+    monteCarloNumBurnInSamples: Int = 100,
+    monteCarloNumSamples: Int = 100,
+    seed: Long = System.currentTimeMillis) {
+
+  /**
+   * Produces a Gaussian Process regression model from the input features and labels
+   *
+   * @param x the observed features
+   * @param y the observed labels
+   * @return the estimated model
+   */
+  def fit(x: DenseMatrix[Double], y: DenseVector[Double]): GaussianProcessModel = {
+    require(x.rows > 0 && x.cols > 0, "Empty input.")
+    require(x.rows == y.length, "Training feature sets and label sets must have the same number of elements")
+
+    // Normalize labels
+    val (yTrain, yMean) = if (normalizeLabels) {
+      val m = mean(y)
+      (y - m, m)
+    } else {
+      (y, 0.0)
+    }
+
+    val kernels = estimateKernelParams(x, yTrain)
+    new GaussianProcessModel(x, yTrain, yMean, kernels, predictionTransformation)
+  }
+
+  /**
+   * Estimates kernel parameters by sampling from the kernel parameter likelihood function
+   *
+   * We assume a uniform prior over the kernel parameters $\theta$ and observed features $x$, therefore:
+   *
+   *   $l(\theta|x,y) = p(y|theta,x) \propto p(theta|x,y)$
+   *
+   * Since the slice sampling algorithm requires that the function be merely proportional to the target distribution,
+   * sampling from this function is equivalent to sampling from p(\theta|x,y). These samples can then be used to compute
+   * a Monte Carlo estimate of the response for a new query point $q'$ by integrating over values of $\theta$:
+   *
+   *   $\int r(x', \theta) p(\theta) d\theta$
+   *
+   * In this way we (approximately) marginalize over all $\theta$ and arrive at a more robust estimate than would be
+   * produced by computing a maximum likelihood point estimate of the parameters.
+   *
+   * @param x the observed features
+   * @param y the observed labels
+   * @return a collection of covariance kernels corresponding to the sampled kernel parameters
+   */
+  protected[estimators] def estimateKernelParams(
+      x: DenseMatrix[Double],
+      y: DenseVector[Double]): List[Kernel] = {
+
+    val logp = (theta: DenseVector[Double]) => logLikelihood(x, y, theta)
+    val sampler = new SliceSampler(logp, range = kernel.getParamBounds, seed = seed)
+
+    // Sampler burn-in. Since Markov chain samplers like slice sampler exhibit serial dependence between samples, the
+    // first n samples are biased by the initial choice of parameter vector. Here we perform a "burn in" procedure to
+    // mitigate this.
+    val init = (0 until monteCarloNumBurnInSamples)
+      .foldLeft(kernel.expandDimensions(kernel.getParams, x.cols)) { (currX, _) =>
+        sampler.draw(currX)
+      }
+
+    // Now draw the real samples from the distribution
+    val (_, samples) = (0 until monteCarloNumSamples)
+      .foldLeft((init, List.empty[DenseVector[Double]])) { case ((currX, ls), _) =>
+        val x = sampler.draw(currX)
+        (x, ls :+ x)
```

Comment:
`currX` isn't the greatest name here. It just gets confusing because we have multiple ideas of what `x` is in the code. Like, here `x` is a point in the Kernel space, then in `GaussianProcessModel` `x` is a vector of several potential hyperparameter tuning vectors, but there's also `xTrain` which is actually the `A` in `Ax = b`.

I can't think of a better name, because that's what it's called in the literature. Maybe a comment to clearly express which context a given `x` falls under? Regardless, I just wanted to put this on the record.